## SQL Practice

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

In [5]:
# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) 
FROM stores;
"""
# this is the exact answer for the asignment
pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:

It has 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * FROM stores ORDER BY name;
"""
pd.read_sql_query(sql, engine)

# This can be the correct asnwer

store                                name store_status  \
0      4986  218 Fuel Express & Chubby's Liquor            A   
1      3047          3047 Cub Foods / Iowa City            I   
2      3060               3060 Cub Foods / Ames            I   
3      3061         3061 Cub Foods / Sioux City            I   
4      4845                  3rd St Convenience            A   
...     ...                                 ...          ...   
1968   3992       Xo Food And Liquor / Waterloo            I   
1969   4066                       Xpress Liquor            I   
1970   4886                      Yr Dollar Stop            A   
1971   4908                Zapf's Pronto Market            A   
1972   4874                      Z's Quickbreak            A   

                                          store_address address_info  
0     68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1     855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2     3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3     1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None  
4     216, 3rd St Se\nCedar Rapids, IA 52401\n(41.97...         None  
...                                                 ...          ...  
1968  428 Franklin St\nWaterloo, IA 507030000\n(42.5...         None  
1969  1201 North Jefferson Ste 100\nIndianola, IA 50...         None  
1970  1500 1st Ave Ne\nCedar Rapids, IA 52402\n(41.9...         None  
1971  107 1st St Sw\nElkader, IA 52043\n(42.85350191...         None  
1972  201, E Washington St\nMount Pleasant, IA 52641...         None  

[1973 rows x 5 columns]

**Prompt 2 Answer**:
3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT * 
FROM sales
WHERE sales."category_name" = 'IMPORTED VODKA'
ORDER BY sales."category_name";
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0      2014-11-13              None   2635            82  Scott  1032080   
1      2014-11-13              None   2635            82  Scott  1032080   
2      2014-11-13              None   2635            82  Scott  1032080   
3      2014-11-13              None   2635            82  Scott  1032080   
4      2014-11-13              None   2635            82  Scott  1032080   
...           ...               ...    ...           ...    ...      ...   
124670 2014-11-13              None   2633            77   Polk  1032080   
124671 2014-11-13              None   2633            77   Polk  1032080   
124672 2014-11-13              None   2633            77   Polk  1032080   
124673 2014-11-13              None   2635            82  Scott  1032080   
124674 2014-11-13              None   2635            82  Scott  1032080   

         category_name vendor_no                            vendor   item  \
0       IMPORTED VODKA       420            Moet Hennessy USA Inc.  34155   
1       IMPORTED VODKA       260                   Diageo Americas  34164   
2       IMPORTED VODKA        85          Brown-Forman Corporation  34326   
3       IMPORTED VODKA        35               Bacardi U.S.A. Inc.  34423   
4       IMPORTED VODKA        35               Bacardi U.S.A. Inc.  34433   
...                ...       ...                               ...    ...   
124670  IMPORTED VODKA       390                  Quadro Group LLC  34678   
124671  IMPORTED VODKA       267              Imperial Brands Inc.  34690   
124672  IMPORTED VODKA       260                   Diageo Americas  36870   
124673  IMPORTED VODKA       370  Pernod Ricard USA/Austin Nichols  34006   
124674  IMPORTED VODKA       370  Pernod Ricard USA/Austin Nichols  34008   

                         description  pack  liter_size state_btl_cost  \
0                    Belvedere Vodka     6         750         $17.15   
1                 Ciroc Luxury Vodka    24         200          $5.00   
2          Finlandia Imp Vodka 80prf    12         750          $8.57   
3                   Grey Goose Vodka    12         375          $9.00   
4                   Grey Goose Vodka    12         750         $18.49   
...                              ...   ...         ...            ...   
124670          Quadro Russian Vodka    12         750         $16.40   
124671                Sobieski Vodka    12         750          $6.50   
124672            Ciroc Luxury Vodka    12        1000         $22.75   
124673  Absolut Swedish Vodka 80 Prf    12         750         $11.49   
124674  Absolut Swedish Vodka 80 Prf     6        1750         $19.99   

       btl_price  bottle_qty   total  
0         $25.73           6  154.38  
1          $7.50           6   45.00  
2         $12.86          12  154.32  
3         $13.50           4   54.00  
4         $27.74          12  332.88  
...          ...         ...     ...  
124670    $24.60           1   24.60  
124671     $9.75          12  117.00  
124672    $34.13          12  409.56  
124673    $17.24          12  206.88  
124674    $29.99          30  899.70  

[124675 rows x 17 columns]

**Prompt 3 Answer**:
 Bottle price with number of bottles  in the order of (\\$ 25.73, 6), (\\$ 7.50, 6) and (\\$ 12.86, 12), respectively

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sales."date", COUNT(total)
FROM sales
WHERE sales."date" BETWEEN '02-01-2015' AND '02-27-2015'
GROUP BY sales."date";
"""
pd.read_sql_query(sql, engine)

date  count
0  2015-02-02   9815
1  2015-02-03   8870
2  2015-02-04  12385
3  2015-02-05   8180
4  2015-02-06     11
5  2015-02-09  10388
6  2015-02-10   8711
7  2015-02-11  12897
8  2015-02-12   7995
9  2015-02-16  10321
10 2015-02-17  10515
11 2015-02-18  11232
12 2015-02-19   8003
13 2015-02-23  10230
14 2015-02-24   9929
15 2015-02-25  10872
16 2015-02-26   7927

**Prompt 4 Answer**:
The table shows the total liquor sold in the whole February 2015

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM((sales."total")*(sales."bottle_qty"))AS "Revenue"
FROM sales
WHERE sales."category_name" = 'IMPORTED VODKA'
ORDER BY "Revenue";
"""

pd.read_sql_query(sql, engine)

Revenue
0  2.057868e+09

**Prompt 5 Answer**:
Revenue value is 2.057868e+09

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM((sales."btl_price")-(sales."state_btl_cost"))AS "Markup"
FROM sales
WHERE sales."category_name" = 'IMPORTED VODKA';

"""
pd.read_sql_query(sql, engine)


Markup
0  $838,671.93

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

**Prompt 7 Answer**:

The % of markup become 0.501169

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT  SUM((sales."btl_price")-(sales."state_btl_cost"))/SUM(sales."state_btl_cost") AS "min_pctMarkup",
        SUM((sales."btl_price")-(sales."state_btl_cost"))/SUM(sales."state_btl_cost") AS "max_pctMarkup"
FROM sales 
LIMIT 5
HAVING MAX(sales."max_pctMarkup") OR MIN(sales."min_pctMarkup");
"""
pd.read_sql_query(sql, engine)


**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT MAX(avg_bottle_sale) AS max_avg_bottle_sale
FROM (
    SELECT county, ROUND(SUM(bottle_qty)/COUNT(county_number),1) AS "avg_bottle_sale"
    FROM county
    INNER JOIN sales ON sales.county = counties.county
    GROUP BY county
) ;
"""

pd.read_sql_query(sql, engine)

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [ ]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) AS "count_total"
FROM (
   SELECT AS sales.*,
        CAST(total AS dollar) AS "conv_total",
        btl_price * bottle_qty AS "cal_total"
   FROM sales
   ) 
WHERE conv_total <> cal_total;
 
"""

pd.read_sql_query(sql, engine)

**Prompt 10 Answer**: